src/Stratis.Bitcoin/Base/InvalidBlockHashStore.cs

```diff
@@ -0,0 +1,154 @@
+    }
+
+    /// <summary>
+    /// In memory store of invalid block header hashes.
+    /// </summary>
+    /// <remarks>
+    /// The store has a limited capacity. When a new block header hash is marked as invalid 
+    /// once the capacity is reached, the oldest entry is removed and no longer considered invalid.
+    /// <para>
+    /// Entries with specified expiration time are either removed just like other entries - i.e. during 
+    /// an add operation when a capacity is reached, or they are removed when they are touched 
+    /// and it is detected that their expiration time is no longer in the future.
+    /// </para>
+    /// </remarks>
+    public class InvalidBlockHashStore : IInvalidBlockHashStore
+    {
+        /// <summary>Default value for the maximal number of hashes we can store.</summary>
+        public const int DefaultCapacity = 1000;
+
+        /// <summary>A provider of the date and time.</summary>
+        private readonly IDateTimeProvider dateTimeProvider;
+
+        /// <summary>Lock object to protect access to <see cref="invalidBlockHashes"/> and <see cref="orderedHashList"/>.</summary>
+        private readonly object lockObject = new object();
+
+        /// <summary>
+        /// Collection of blocks that are to be considered invalid. If the value of the entry is not <c>null</c>, 
+        /// the entry is considered invalid only for a certain amount of time.
+        /// </summary>
+        /// <remarks>All access to this object has to be protected by <see cref="lockObject"/>.</remarks>
+        private readonly Dictionary<uint256, DateTime?> invalidBlockHashes;
+
+        /// <summary>Circular array of block header hash entries to allow quick removal of the oldest entry once the capacity is reached.</summary>
+        /// <remarks>
+        /// All access to this object has to be protected by <see cref="lockObject"/>.
+        /// <para>The field is internal for testing purposes.</para>
+        /// </remarks>
+        internal readonly CircularArray<uint256> orderedHashList;
+
+        /// <summary>
+        /// Initializes the instance of the object.
+        /// </summary>
+        /// <param name="capacity">Maximal number of hashes we can store.</param>
+        public InvalidBlockHashStore(IDateTimeProvider dateTimeProvider, int capacity = DefaultCapacity)
+        {
+            this.dateTimeProvider = dateTimeProvider;
+
+            this.invalidBlockHashes = new Dictionary<uint256, DateTime?>(capacity);
+            this.orderedHashList = new CircularArray<uint256>(capacity);
+        }
+
+        /// <inheritdoc />
+        public bool IsInvalid(uint256 hashBlock)
+        {
+            bool res = false;
+
+            lock (this.lockObject)
+            {
+                // First check if the entry exists.
+                DateTime? expirationTime;
+                if (this.invalidBlockHashes.TryGetValue(hashBlock, out expirationTime))
```

Comment:
I think that the name "invalidBlockHashes" is confusing when used in this context. I.e you are calling a TryGetValue on a field called "invalidBlockHashes" but returning a date? Can we maybe rename this to something more descriptive like invalidBlockHashesByExpiry or invalidBlockExpiryHashes ?

Reply:
you are right, the current name is not perfect

but it is actually more like invalidBlockHashesExpirationTimesByBlockHashes or "expirations-by-hashes", rather than "hashes-by-expiration", so the other way around, so i am trying to find a sensible name for it ...

maybe invalidBlockHashesData or invalidBlockHashesInfo, what do you think?

Reply:
Hmmmm what about invalidBlockExpirationsByHash? Then we know we return an expiry when we TryGetValue?

Reply:
i am thinking about this, this is a classic example and i can see where the confusion is coming from

in general, this is classic situation when you have ID and RECORD, so we have Dictionary(ID, RECORD) and RECORD is 
```
public class RECORD
{
  public Type1 ID;
  public Type2 Data1;
  public Type3 Data2;
  public Type4 Data3;
  ...
}

and we would name such a things as "private Dict<ID, RECORD> RECORD records;"
```

and in this case ID is uint256 and is actually the most important thing in the record
and the only other data we have for it is that `DateTime?`

so instead of creating this class holding the whole record information, we take a shortcut and say that value in the dict is just `DateTime?`, but in fact what we want is the whole structure `{ uint256, DateTime? }` and we have it when we work with the dict

the hash here is the primary purpose, DateTime is secondary (but still very important), this is why i don't really like invalidBlockExpirationsByHash because the pattern is invalidRECORD-By-ID
and as explained, it is not the DateTime that is the whole record here

so this is why I am more for "invalidBlockHashesData" or "invalidBlockHashesInfo"

Reply:
and that's why the current name is not correct as you pointed out correctly, because the RECORD here is not just the hash, so i very much agree with renaming, just not sure to what

@dangershony sometimes have good ideas for names, could he help possibly?

my best candidates at this moment are  "invalidBlockHashesData" or "invalidBlockHashesInfo"

Reply:
How about we introduce a small class that holds that data? Something called InvalidBlockHashExpiry which holds the hash as well as the expiry date?

```
public class InvalidBlockHashExpiry 
{
    public uint BlockHash {get;set;}
    public DateTime Expiry {get;set;}
}
```

Then you have a list instead of a dictionary?

Or will this be too much overhead?

Reply:
actually the whole trouble was to find the name for such a class, because that is immediately the name of the dictionary

dictionary is used for O(1) access, so that should stay, 

but `invalidBlockHashExpirations` is probably a good name for the dictionary, are you happy with that?

Reply:
I'm happy with ```invalidBlockHashExpirations``` @Aprogiena 👍

Reply:
wowoo I was fine with any name to be honest as this whole class is for expired blocks. 
My favourite was `invalidBlockHashesByExpiry` but no big deal.